### PR TITLE
Add a mixin to get SPIP version and make use of it

### PR DESCRIPTION
**Pull request by jvoisin**
This was sparked by #19394

### CODE REVIEW

**Comment by dwelch-r7** [comment]
I've just merged in https://github.com/rapid7/metasploit-framework/pull/19394 would you like to update this PR to include this mixin in the new module too? or I can do it totally up to you

**Comment by jvoisin** [comment]
Feel free to do it, it'll be faster than another review round-trip :)

File: lib/msf/core/exploit/remote/http/spip.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      'uri'    => normalize_uri(target_uri.path, "spip.php")
+    )
+
+    return unless res
+    
+    version = res.get_html_document.at('head/meta[@name="generator"]/@content')&.text
+
+    if version.nil? && res.headers['Composed-By'] =~ /SPIP (.*) @/
+      version = ::Regexp.last_match(1)
+    end
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/remote/http/spip.rb:31]: I see you're pulling the version from two different places here. Is there any context you can easily add in a comment about when one technique should work versus the other? What I'm wondering is if there's a version where prior one technique is used and after another is used or if it's dependent on some kind of configuration setting.

**Review by smcintyre-r7 — COMMENTED** [review]
I just left one question about the version fingerprinting. This code all looks great, so thanks for making it into a reusable library @jvoisin.

File: lib/msf/core/exploit/remote/http/spip.rb (hunk 1 above)
Comment (jvoisin) [reply lib/msf/core/exploit/remote/http/spip.rb:31]: I think it depends on the configuration/version of Spip, but I don't know the details. Maybe @laluka knows?

Comment (laluka) [reply lib/msf/core/exploit/remote/http/spip.rb:31]: Henlo !
It's all depends on the config, whether files are accessible, headerd, or reflected html content, there's sadly no thumb rule :/

Nice addon btw, and have a nice day! 🌹

File: modules/exploits/unix/webapp/spip_rce_form.rb [hunk 2]
```diff
@@ -83,38 +84,20 @@ def check
     res = send_request_cgi({ 'uri' => uri.to_s })
 
     return Exploit::CheckCode::Unknown('Target is unreachable.') unless res
-    return Exploit::CheckCode::Unknown("Target responded with unexpected HTTP response code: #{res.code}") unless res.code == 200
 
-    version_string = res.get_html_document.at('head/meta[@name="generator"]/@content')&.text
-    return Exploit::CheckCode::Unknown('Unable to find the version string on the page: spip.php') unless version_string =~ /SPIP (.*)/
+    rversion = spip_version
```
Comment (dwelch-r7) [inline modules/exploits/unix/webapp/spip_rce_form.rb:88]:
```
msf6 exploit(unix/webapp/spip_rce_form) > check http://127.0.0.1:8000

[-] Exploit failed: ArgumentError Malformed version number string SPIP 4.2.12
[-] 127.0.0.1:8000 - Check failed: The state could not be determined.
```
Seems like this is pulling out `SPIP 4.2.12` for me in this module and Rex:Version does not like that
I replicated on the other module too, not sure if this is a version specific thing or not

Comment (jvoisin) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]: You're absolutely right, I didn't test the PR well enough. I just pushed some changes that should fix this.

Comment (dwelch-r7) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]:
```
msf6 exploit(unix/webapp/spip_rce_form) > check http://127.0.0.1:8000

[-] Exploit failed: ArgumentError Malformed version number string 4.2.12 @ www.spip.net +
[-] 127.0.0.1:8000 - Check failed: The state could not be determined.
```
getting a slightly different error now

Comment (jvoisin) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]: I'm confused as why the regex (`/SPIP (.*) /`) doesn't match. I don't have a testing environment handy (holidays, yay!), but something like `/SPIP ([0-9.]+) /` might fix the issue?

Comment (dwelch-r7) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]: I think we want `/SPIP (.*)/` seems like there was a rogue space messing things up
```
[7] pry(#<Msf::Modules::Exploit__Unix__Webapp__Spip_rce_form::MetasploitModule>)> version_string
=> "SPIP 4.2.12"
[8] pry(#<Msf::Modules::Exploit__Unix__Webapp__Spip_rce_form::MetasploitModule>)> version_string =~ /SPIP (.*) /
=> nil
[9] pry(#<Msf::Modules::Exploit__Unix__Webapp__Spip_rce_form::MetasploitModule>)> version_string =~ /SPIP (.*)/
=> 0
[10] pry(#<Msf::Modules::Exploit__Unix__Webapp__Spip_rce_form::MetasploitModule>)> ::Regexp.last_match(1)
=> "4.2.12"
```

Comment (jvoisin) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]: Nice catch! Care to fix it so this can land, or do you prefer to wait until I get back home?

Comment (dwelch-r7) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]: I can do it no problem

Comment (jvoisin) [reply modules/exploits/unix/webapp/spip_rce_form.rb:88]: I'll owe your a beverage of your choice shall we ever meet in the meatspace then <3

**Review by cgranleese-r7 — APPROVED** [review]

**Comment by dwelch-r7** [comment]
# Release Notes

Adda a mixin to get SPIP version and make use of it

**Pull request merged**